pkg/controllers/corecluster/corecluster_controller_test.go

```diff
@@ -0,0 +1,244 @@
+	capabuilder "github.com/openshift/cluster-api-actuator-pkg/testutils/resourcebuilder/cluster-api/infrastructure/v1beta2"
+	configv1resourcebuilder "github.com/openshift/cluster-api-actuator-pkg/testutils/resourcebuilder/config/v1"
+	corev1resourcebuilder "github.com/openshift/cluster-api-actuator-pkg/testutils/resourcebuilder/core/v1"
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+)
+
+var _ = Describe("Reconcile Core cluster", func() {
+	var r *CoreClusterReconciler
+	var coreCluster *capiv1.Cluster
+	var testNamespaceName string
+	testInfraName := "test-ocp-infrastructure-name"
+	testRegionName := "eu-west-2"
+	var ocpInfrastructure *configv1.Infrastructure
+	var ocpInfrastructureCopy *configv1.Infrastructure
+
+	Context("With a supported platform", func() {
+		BeforeEach(func() {
+			By("creating the testing namespace")
+			namespace := corev1resourcebuilder.Namespace().WithGenerateName("test-capi-corecluster-").Build()
+			Expect(cl.Create(context.Background(), namespace)).To(Succeed())
+			testNamespaceName = namespace.Name
+
+			//TODO: implement unit tests across multiple providers
+			By("creating the testing OCP infrastructure on AWS")
+			ocpInfrastructure = configv1resourcebuilder.Infrastructure().AsAWS(testInfraName, testRegionName).Build()
+			Expect(cl.Create(ctx, ocpInfrastructure)).To(Succeed())
+
+			ocpInfrastructureCopy = ocpInfrastructure.DeepCopy()
+			ocpInfrastructureCopy.Status = configv1.InfrastructureStatus{
+				APIServerInternalURL: "https://test:8081",
+				InfrastructureName:   testInfraName,
+				Platform:             configv1.AWSPlatformType,
+				PlatformStatus: &configv1.PlatformStatus{
+					Type: configv1.AWSPlatformType,
+					AWS: &configv1.AWSPlatformStatus{
+						Region: testRegionName,
+					},
+				},
+			}
+			Expect(cl.Status().Patch(ctx, ocpInfrastructure, client.MergeFrom(ocpInfrastructureCopy))).To(Succeed())
+		})
+
+		AfterEach(func() {
+			testutils.CleanupResources(Default, ctx, testEnv.Config, cl, testNamespaceName, &configv1.Infrastructure{}, &capiv1.Cluster{}, &capav1.AWSCluster{})
+		})
+
+		Context("When there is no Core Cluster", func() {
+			BeforeEach(func() {
+				r = &CoreClusterReconciler{
+					ClusterOperatorStatusClient: operatorstatus.ClusterOperatorStatusClient{
+						Client: cl,
+					},
+					Cluster:  &capiv1.Cluster{},
+					Infra:    ocpInfrastructureCopy.DeepCopy(),
+					Platform: ocpInfrastructureCopy.Status.PlatformStatus.Type,
+				}
+			})
+
+			Context("When there is no Infra Cluster", func() {
+				It("should error saying there is no infra cluster", func() {
+					By("running the core cluster controller reconciliation")
+					_, err := r.Reconcile(ctx, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Namespace: testNamespaceName,
+							Name:      testInfraName,
+						},
+					})
+					Expect(err).To(HaveOccurred()) // TODO match error.
```

Comment:
Can we fix error matching before we merge?

Reply:
Done :) (for now, have to add a test or two more)